xc/doc/man/Xv/Xv.man
.TH Xv 3X
.\"
.\" $XFree86: xc/doc/man/Xv/Xv.man,v 1.4 1999/03/02 11:49:13 dawes Exp $
.SH Name
Xv --- X Window System video extension
.SH Description
The Xv extension provides support for video adaptors attached to an X
display.  It takes the approach that a display may have one or more
video adaptors, each of which has one or more ports through which
independent video streams pass.  
.PP
An adaptor may be able to display video in a drawable, capture video
from a drawable, or both.  It translates between video encoding (NTSC,
PAL, SECAM, etc...) and drawable format (depth and visual-id pair). An
adaptor may support multiple video encodings and/or multiple drawable
formats.
.PP
Clients use Xv(3X) to gain access and manage sharing of a display's
video resources.  Typically, a client will use XvQueryExtension(3X) to
determine the status of the extension, XvQueryAdaptors(3X) to get a
description of what video adaptors exist, and XvQueryEncodings(3X) to
get a description of what video encodings an adaptor supports.
.PP
Once a client has determined what video resources are available, it is
free to put video into a drawable or get video from a drawable,
according the capabilities supported.  Clients can select to receive
events when video activity changes in a drawable and when port
attributes have changed.
.SH Summary of Library Functions
The following is a summary of Xv library functions:
.sp 1l
XvGetPortAttribute(3X) \- return current port attribute value
.sp 1l
XvGetStill(3X) \- capture a single frame of video from a drawable
.sp 1l
XvGetVideo(3X) \- capture video from a drawable
.sp 1l
XvGrabPort(3X) \- lock port for exclusive use by client
.sp 1l
XvPortNotify(3X) \- event generated when port attributes change
.sp 1l
XvPutStill(3X) \- write a single frame of video to a drawable
.sp 1l
XvPutVideo(3X) \- write video into a drawable
.sp 1l
XvQueryAdaptors(3X) \- return adaptor information for a screen
.sp 1l
XvQueryBestSize(3X) \- determine the optimum drawable region size
.sp 1l
XvQueryEncodings(3X) \- return list of encodings for an adaptor
.sp 1l
XvQueryExtension(3X) \- return version and revision of extension
.sp 1l
XvSelectPortNotify(3X) \- enable or disable XvPortNotify events
.sp 1l
XvSelectVideoNotify(3X) \- enable or disable XvVideoNotify events
.sp 1l
XvSetPortAttribute(3X) \- set an attribute for a port 
.sp 1l
XvStopVideo(3X) \- stop active video 
.sp 1l
XvUngrabPort(3X) \- release a grabbed port
.sp 1l
XvVideoNotify(3X) \- event generated for video processing
.sp 1l
.PP
Each of these functions is described in its own Xv man page.
.br



xc/lib/Xv/Xv.c
/***********************************************************
Copyright 1991 by Digital Equipment Corporation, Maynard, Massachusetts,
and the Massachusetts Institute of Technology, Cambridge, Massachusetts.

                        All Rights Reserved

Permission to use, copy, modify, and distribute this software and its 
documentation for any purpose and without fee is hereby granted, 
provided that the above copyright notice appear in all copies and that
both that copyright notice and this permission notice appear in 
supporting documentation, and that the names of Digital or MIT not be
used in advertising or publicity pertaining to distribution of the
software without specific, written prior permission.  

DIGITAL DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING
ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS, IN NO EVENT SHALL
DIGITAL BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR
ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,
WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION,
ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS
SOFTWARE.

******************************************************************/
/* $XFree86: xc/lib/Xv/Xv.c,v 1.12 2000/03/03 20:51:09 mvojkovi Exp $ */
/*
** File: 
**
**   Xv.c --- Xv library extension module.
**
** Author: 
**
**   David Carver (Digital Workstation Engineering/Project Athena)
**
** Revisions:
**
**   26.06.91 Carver
**     - changed XvFreeAdaptors to XvFreeAdaptorInfo
**     - changed XvFreeEncodings to XvFreeEncodingInfo
**
**   11.06.91 Carver
**     - changed SetPortControl to SetPortAttribute
**     - changed GetPortControl to GetPortAttribute
**     - changed QueryBestSize
**
**   15.05.91 Carver
**     - version 2.0 upgrade
**
**   240.01.91 Carver
**     - version 1.4 upgrade
**
*/

#include "Xvlibint.h"
#include "extutil.h"
#include "XShm.h"

static XExtCodes *_XvCodes;
static Bool (* _XvOldWireToEventVideo)();
static Bool (* _XvOldWireToEventPort)();
static char *(* _XvOldErrorString)();

static Bool _XvWireToEvent();
static char *_XvErrorString();

#define	PREAMBLE(stat) \
  LockDisplay(dpy); \
  if ((!_XvCodes) && (!_XvInitExtension(dpy))) \
    { \
      UnlockDisplay(dpy); SyncHandle(); return (stat); \
    }

#define POSTAMBLE \
  UnlockDisplay(dpy); \
  SyncHandle()

int
XvQueryExtension(
     Display *dpy,
     unsigned int *p_version,
     unsigned int *p_revision,
     unsigned int *p_requestBase,
     unsigned int *p_eventBase,
     unsigned int *p_errorBase
){
  xvQueryExtensionReq *req;
  xvQueryExtensionReply rep;

  PREAMBLE(XvBadExtension);

  XvGetReq(QueryExtension, req);

  /* READ THE REPLY */

  if (!_XReply(dpy, (xReply *)&rep, 0, xFalse)) 
    {
      POSTAMBLE;
      return(XvBadExtension);
    }

  *p_version = rep.version;
  *p_revision = rep.revision;
  *p_requestBase = _XvCodes->major_opcode;
  *p_eventBase = _XvCodes->first_event;
  *p_errorBase = _XvCodes->first_error;

  POSTAMBLE;

  return Success;
}

int
XvQueryAdaptors(
     Display *dpy,
     Window window,
     unsigned int *p_nAdaptors,
     XvAdaptorInfo **p_pAdaptors
){
  xvQueryAdaptorsReq *req;
  xvQueryAdaptorsReply rep;
  int size,ii,jj;
  char *name;
  XvAdaptorInfo *pas, *pa;
  XvFormat *pfs, *pf;
  char *buffer;
  union 
    {
      char *buffer;
      char *string;
      xvAdaptorInfo *pa;
      xvFormat *pf;
    } u;
  
  PREAMBLE(XvBadExtension);

  XvGetReq(QueryAdaptors, req);
  req->window = window;

  /* READ THE REPLY */

  if (_XReply(dpy, (xReply *)&rep, 0, xFalse) == 0) 
    {
      POSTAMBLE;
      return(XvBadReply);
    }

  size = rep.length << 2;
  if ( (buffer = (char *)Xmalloc ((unsigned) size)) == NULL)
    {
      POSTAMBLE;
      return(XvBadAlloc);
    }
  _XRead (dpy, buffer, size);

  u.buffer = buffer;

  /* GET INPUT ADAPTORS */

  size = rep.num_adaptors*sizeof(XvAdaptorInfo);
  if ((pas=(XvAdaptorInfo *)Xmalloc(size))==NULL)
    {
      Xfree(buffer);
      POSTAMBLE;
      return(XvBadAlloc);
    }

  /* INIT ADAPTOR FIELDS */

  pa = pas;
  for (ii=0; ii<rep.num_adaptors; ii++)
    {
      pa->num_adaptors = 0;
      pa->name = (char *)NULL;
      pa->formats = (XvFormat *)NULL;
    }

  pa = pas;
  for (ii=0; ii<rep.num_adaptors; ii++)
    {
      pa->type = u.pa->type;
      pa->base_id = u.pa->base_id;
      pa->num_ports = u.pa->num_ports;
      pa->num_formats = u.pa->num_formats;
      pa->num_adaptors = rep.num_adaptors - ii;

      /* GET ADAPTOR NAME */

      size = u.pa->name_size;
      u.buffer += (sz_xvAdaptorInfo + 3) & ~3;

      if ( (name = (char *)Xmalloc(size+1)) == NULL)
	{
	  XvFreeAdaptorInfo(pas);
	  Xfree(buffer);
	  POSTAMBLE;
	  return(XvBadAlloc);
	}
      (void)strncpy(name, u.string, size);
      name[size] = '\0';
      pa->name = name;

      u.buffer += (size + 3) & ~3;

      /* GET FORMATS */

      size = pa->num_formats*sizeof(XvFormat);
      if ((pfs=(XvFormat *)Xmalloc(size))==NULL)
	{
	  XvFreeAdaptorInfo(pas);
	  Xfree(buffer);
	  POSTAMBLE;
	  return(XvBadAlloc);
	}

      pf = pfs;
      for (jj=0; jj<pa->num_formats; jj++)
	{
	  pf->depth = u.pf->depth;
	  pf->visual_id = u.pf->visual;
	  pf++;
	  
	  u.buffer += (sz_xvFormat + 3) & ~3;
	}

      pa->formats = pfs;

      pa++;

    }

  *p_nAdaptors = rep.num_adaptors;
  *p_pAdaptors = pas;

  POSTAMBLE;

  return (Success);
}


void
XvFreeAdaptorInfo(XvAdaptorInfo *pAdaptors)
{

  XvAdaptorInfo *pa;
  int ii;

  if (!pAdaptors) return;

  pa = pAdaptors;

  for (ii=0; ii<pAdaptors->num_adaptors; ii++, pa++)
    {
      if (pa->name)
	{
	  Xfree(pa->name);
	}
      if (pa->formats)
	{
	  Xfree(pa->formats);
	}
    } 

  Xfree(pAdaptors);
}

int
XvQueryEncodings(
     Display *dpy,
     XvPortID port,
     unsigned int *p_nEncodings,
     XvEncodingInfo **p_pEncodings
){
  xvQueryEncodingsReq *req;
  xvQueryEncodingsReply rep;
  int size, jj;
  char *name;
  XvEncodingInfo *pes, *pe;
  char *buffer;
  union 
    {
      char *buffer;
      char *string;
      xvEncodingInfo *pe;
    } u;
  
  PREAMBLE(XvBadExtension);

  XvGetReq(QueryEncodings, req);
  req->port = port;

  /* READ THE REPLY */

  if (_XReply(dpy, (xReply *)&rep, 0, xFalse) == 0) 
    {
      POSTAMBLE;
      return(XvBadReply);
    }

  size = rep.length << 2;
  if ( (buffer = (char *)Xmalloc ((unsigned) size)) == NULL)
    {
      POSTAMBLE;
      return(XvBadAlloc);
    }
  _XRead (dpy, buffer, size);

  u.buffer = buffer;

  /* GET ENCODINGS */

  size = rep.num_encodings*sizeof(XvEncodingInfo);
  if ( (pes = (XvEncodingInfo *)Xmalloc(size)) == NULL)
    {
      Xfree(buffer);
      POSTAMBLE;
      return(XvBadAlloc);
    }

  /* INITIALIZE THE ENCODING POINTER */

  pe = pes;
  for (jj=0; jj<rep.num_encodings; jj++)
    {
      pe->name = (char *)NULL;
      pe->num_encodings = 0;
      pe++;
    }

  pe = pes;
  for (jj=0; jj<rep.num_encodings; jj++)
    {
      pe->encoding_id = u.pe->encoding;
      pe->width = u.pe->width;
      pe->height = u.pe->height;
      pe->rate.numerator = u.pe->rate.numerator;
      pe->rate.denominator = u.pe->rate.denominator;
      pe->num_encodings = rep.num_encodings - jj;

      size = u.pe->name_size;
      u.buffer += (sz_xvEncodingInfo + 3) & ~3;

      if ( (name = (char *)Xmalloc(size+1)) == NULL)
	{
	  Xfree(buffer);
	  POSTAMBLE;
	  return(XvBadAlloc);
	}
      strncpy(name, u.string, size);
      name[size] = '\0';
      pe->name = name;
      pe++;

      u.buffer += (size + 3) & ~3;

    }

  *p_nEncodings = rep.num_encodings;
  *p_pEncodings = pes;

  POSTAMBLE;

  return (Success);
}

void
XvFreeEncodingInfo(XvEncodingInfo *pEncodings)
{

  XvEncodingInfo *pe;
  int ii;

  if (!pEncodings) return;

  pe = pEncodings;

  for (ii=0; ii<pEncodings->num_encodings; ii++, pe++)
    {
      if (pe->name) Xfree(pe->name);
    }

  Xfree(pEncodings);

}

int
XvPutVideo(
     Display *dpy,
     XvPortID port,
     Drawable d,
     GC gc,
     int vx, int vy, 
     unsigned int vw, unsigned int vh,
     int dx, int dy,
     unsigned int dw, unsigned int dh
){
  xvPutVideoReq *req;

  PREAMBLE(XvBadExtension);
  
  FlushGC(dpy, gc);

  XvGetReq(PutVideo, req);

  req->port = port;
  req->drawable = d;
  req->gc = gc->gid;
  req->vid_x = vx;
  req->vid_y = vy;
  req->vid_w = vw;
  req->vid_h = vh;
  req->drw_x = dx;
  req->drw_y = dy;
  req->drw_w = dw;
  req->drw_h = dh;

  POSTAMBLE;

  return Success;
}

int
XvPutStill(
     Display *dpy,
     XvPortID port,
     Drawable d,
     GC gc,
     int vx, int vy, 
     unsigned int vw, unsigned int vh,
     int dx, int dy,
     unsigned int dw, unsigned int dh
){
  xvPutStillReq *req;

  PREAMBLE(XvBadExtension);

  FlushGC(dpy, gc);

  XvGetReq(PutStill, req);
  req->port = port;
  req->drawable = d;
  req->gc = gc->gid;
  req->vid_x = vx;
  req->vid_y = vy;
  req->vid_w = vw;
  req->vid_h = vh;
  req->drw_x = dx;
  req->drw_y = dy;
  req->drw_w = dw;
  req->drw_h = dh;

  POSTAMBLE;

  return Success;
}

int
XvGetVideo(
     Display *dpy,
     XvPortID port,
     Drawable d,
     GC gc,
     int vx, int vy, 
     unsigned int vw, unsigned int vh,
     int dx, int dy,
     unsigned int dw, unsigned int dh
){
  xvGetVideoReq *req;

  PREAMBLE(XvBadExtension);

  FlushGC(dpy, gc);

  XvGetReq(GetVideo, req);
  req->port = port;
  req->drawable = d;
  req->gc = gc->gid;
  req->vid_x = vx;
  req->vid_y = vy;
  req->vid_w = vw;
  req->vid_h = vh;
  req->drw_x = dx;
  req->drw_y = dy;
  req->drw_w = dw;
  req->drw_h = dh;

  POSTAMBLE;

  return Success;
}

int
XvGetStill(
     Display *dpy,
     XvPortID port,
     Drawable d,
     GC gc,
     int vx, int vy, 
     unsigned int vw, unsigned int vh,
     int dx, int dy,
     unsigned int dw, unsigned int dh
){
  xvGetStillReq *req;

  PREAMBLE(XvBadExtension);

  FlushGC(dpy, gc);

  XvGetReq(GetStill, req);
  req->port = port;
  req->drawable = d;
  req->gc = gc->gid;
  req->vid_x = vx;
  req->vid_y = vy;
  req->vid_w = vw;
  req->vid_h = vh;
  req->drw_x = dx;
  req->drw_y = dy;
  req->drw_w = dw;
  req->drw_h = dh;

  POSTAMBLE;

  return Success;
}

int
XvStopVideo(
     Display *dpy,
     XvPortID port,
     Drawable draw
){
  xvStopVideoReq *req;

  PREAMBLE(XvBadExtension);

  XvGetReq(StopVideo, req);
  req->port = port;
  req->drawable = draw;

  POSTAMBLE;

  return Success;
}

int
XvGrabPort(
     Display *dpy,
     XvPortID port,
     Time time
){
  int result;
  xvGrabPortReply rep;
  xvGrabPortReq *req;

  PREAMBLE(XvBadExtension);

  XvGetReq(GrabPort, req);
  req->port = port;
  req->time = time;

  if (_XReply (dpy, (xReply *) &rep, 0, xTrue) == 0) 
    rep.result = GrabSuccess;

  result = rep.result;

  POSTAMBLE;

  return result;
}

int
XvUngrabPort(
     Display *dpy,
     XvPortID port,
     Time time
){
  xvUngrabPortReq *req;

  PREAMBLE(XvBadExtension);

  XvGetReq(UngrabPort, req);
  req->port = port;
  req->time = time;

  POSTAMBLE;

  return Success;
}

int
XvSelectVideoNotify(
     Display *dpy,
     Drawable drawable,
     Bool onoff
){
  register xvSelectVideoNotifyReq *req;

  PREAMBLE(XvBadExtension);

  XvGetReq(SelectVideoNotify, req);
  req->drawable = drawable;
  req->onoff = onoff;

  POSTAMBLE;

  return Success;
}

int
XvSelectPortNotify(
     Display *dpy,
     XvPortID port,
     Bool onoff
){
  xvSelectPortNotifyReq *req;

  PREAMBLE(XvBadExtension);

  XvGetReq(SelectPortNotify, req);
  req->port = port;
  req->onoff = onoff;

  POSTAMBLE;

  return Success;
}

int
XvSetPortAttribute (dpy, port, attribute, value)
     register Display *dpy;
     XvPortID port;
     Atom attribute;
     int value;
{
  register xvSetPortAttributeReq *req;

  PREAMBLE(XvBadExtension);

  XvGetReq(SetPortAttribute, req);
  req->port = port;
  req->attribute = attribute;
  req->value = value;

  POSTAMBLE;

  return (Success);
}

int
XvGetPortAttribute (dpy, port, attribute, p_value)
     register Display *dpy;
     XvPortID port;
     Atom attribute;
     int *p_value;
{
  register xvGetPortAttributeReq *req;
  xvGetPortAttributeReply rep;

  PREAMBLE(XvBadExtension);

  XvGetReq(GetPortAttribute, req);
  req->port = port;
  req->attribute = attribute;

  /* READ THE REPLY */

  if (_XReply(dpy, (xReply *)&rep, 0, xFalse) == 0) 
    {
      POSTAMBLE;
      return(XvBadReply);
    }

  *p_value = rep.value;
  
  POSTAMBLE;

  return (Success);
}

int
XvQueryBestSize(dpy, port, motion, vid_w, vid_h, drw_w, drw_h, 
		p_actual_width, p_actual_height)
     register Display *dpy;
     XvPortID port;
     Bool motion;
     unsigned int vid_w, vid_h;
     unsigned int drw_w, drw_h;
     unsigned int *p_actual_width, *p_actual_height;
{
  register xvQueryBestSizeReq *req;
  xvQueryBestSizeReply rep;

  PREAMBLE(XvBadExtension);

  XvGetReq(QueryBestSize, req);
  req->port = port;
  req->motion = motion;
  req->vid_w = vid_w;
  req->vid_h = vid_h;
  req->drw_w = drw_w;
  req->drw_h = drw_h;

  /* READ THE REPLY */

  if (_XReply(dpy, (xReply *)&rep, 0, xFalse) == 0) 
    {
      POSTAMBLE;
      return(XvBadReply);
    }

  *p_actual_width = rep.actual_width;
  *p_actual_height = rep.actual_height;

  POSTAMBLE;

  return (Success);
}


XvAttribute* 
XvQueryPortAttributes(Display *dpy, XvPortID port, int *num)
{
  xvQueryPortAttributesReq *req;
  xvQueryPortAttributesReply rep;
  XvAttribute *ret = NULL;

  *num = 0;

  PREAMBLE(NULL);

  XvGetReq(QueryPortAttributes, req);
  req->port = port;

  /* READ THE REPLY */

  if (_XReply(dpy, (xReply *)&rep, 0, xFalse) == 0) {
      POSTAMBLE;
      return ret;
  }

  if(rep.num_attributes) {
      int size = (rep.num_attributes * sizeof(XvAttribute)) + rep.text_size;

      if((ret = Xmalloc(size))) {
	  char* marker = (char*)(&ret[rep.num_attributes]);
	  xvAttributeInfo Info;
	  int i;
	
	  for(i = 0; i < rep.num_attributes; i++) {
             _XRead(dpy, (char*)(&Info), sz_xvAttributeInfo);
	      ret[i].flags = (int)Info.flags;	      
	      ret[i].min_value = Info.min;	      
	      ret[i].max_value = Info.max;	      
	      ret[i].name = marker;
	      _XRead(dpy, marker, Info.size);
	      marker += Info.size;
	      (*num)++;
	  }
      } else
	_XEatData(dpy, rep.length << 2);
  }

  POSTAMBLE;

  return ret;
}

XvImageFormatValues * XvListImageFormats (
   Display 	*dpy,
   XvPortID 	port,
   int 		*num
){
  xvListImageFormatsReq *req;
  xvListImageFormatsReply rep;
  XvImageFormatValues *ret = NULL;

  *num = 0;

  PREAMBLE(NULL);

  XvGetReq(ListImageFormats, req);
  req->port = port;

  /* READ THE REPLY */

  if (_XReply(dpy, (xReply *)&rep, 0, xFalse) == 0) {
      POSTAMBLE;
      return NULL;
  }

  if(rep.num_formats) {
      int size = (rep.num_formats * sizeof(XvImageFormatValues));

      if((ret = Xmalloc(size))) {
	  xvImageFormatInfo Info;
	  int i;
	
	  for(i = 0; i < rep.num_formats; i++) {
              _XRead(dpy, (char*)(&Info), sz_xvImageFormatInfo);
	      ret[i].id = Info.id;	      
	      ret[i].type = Info.type;	      
	      ret[i].byte_order = Info.byte_order;	      
	      memcpy(&(ret[i].guid[0]), &(Info.guid[0]), 16);
	      ret[i].bits_per_pixel = Info.bpp;	      
  	      ret[i].format = Info.format;	      
   	      ret[i].num_planes = Info.num_planes;	      
    	      ret[i].depth = Info.depth;	      
    	      ret[i].red_mask = Info.red_mask;	      
    	      ret[i].green_mask = Info.green_mask;	      
    	      ret[i].blue_mask = Info.blue_mask;	      
    	      ret[i].y_sample_bits = Info.y_sample_bits;	      
    	      ret[i].u_sample_bits = Info.u_sample_bits;	      
    	      ret[i].v_sample_bits = Info.v_sample_bits;
    	      ret[i].horz_y_period = Info.horz_y_period;
    	      ret[i].horz_u_period = Info.horz_u_period;
    	      ret[i].horz_v_period = Info.horz_v_period;
    	      ret[i].vert_y_period = Info.vert_y_period;
    	      ret[i].vert_u_period = Info.vert_u_period;
    	      ret[i].vert_v_period = Info.vert_v_period;
	      memcpy(&(ret[i].component_order[0]), &(Info.comp_order[0]), 32);
    	      ret[i].scanline_order = Info.scanline_order;
	      (*num)++;
	  }
      } else
	_XEatData(dpy, rep.length << 2);
  }

  POSTAMBLE;

  return ret;
}

XvImage * XvCreateImage (
   Display *dpy,
   XvPortID port,
   int id,
   char *data,
   int width, 
   int height 
) {
   xvQueryImageAttributesReq *req;
   xvQueryImageAttributesReply rep;
   XvImage *ret = NULL;

   PREAMBLE(NULL);

   XvGetReq(QueryImageAttributes, req);
   req->id = id;
   req->port = port;
   req->width = width;
   req->height = height;

   /* READ THE REPLY */

   if (!_XReply(dpy, (xReply *)&rep, 0, xFalse)) {
      POSTAMBLE;
      return NULL;
   }

   if(ret = (XvImage*)Xmalloc(sizeof(XvImage) + (rep.num_planes << 3))) {
	ret->id = id;
	ret->width = rep.width;
	ret->height = rep.height;
	ret->data_size = rep.data_size;
	ret->num_planes = rep.num_planes;
	ret->pitches = (int*)(&ret[1]);
	ret->offsets = ret->pitches + rep.num_planes;
	ret->data = data;
	ret->obdata = NULL;
  	_XRead(dpy, (char*)(ret->pitches), rep.num_planes << 2);
	_XRead(dpy, (char*)(ret->offsets), rep.num_planes << 2);
   } else
	_XEatData(dpy, rep.length << 2);

   POSTAMBLE;

   return ret;
}

XvImage * XvShmCreateImage (
   Display *dpy,
   XvPortID port,
   int id,
   char *data,
   int width, 
   int height,
   XShmSegmentInfo *shminfo
){
   XvImage *ret;

   ret = XvCreateImage(dpy, port, id, data, width, height);

   if(ret) ret->obdata = (XPointer)shminfo;

   return ret;
}

int XvPutImage (
   Display *dpy,
   XvPortID port,
   Drawable d,
   GC gc,
   XvImage *image,
   int src_x,
   int src_y,
   unsigned int src_w,
   unsigned int src_h,
   int dest_x, 
   int dest_y,
   unsigned int dest_w,
   unsigned int dest_h
){
  xvPutImageReq *req;

  PREAMBLE(XvBadExtension);

  if(image->data_size > (dpy->max_request_size << 2)) /* FixMe */
	return BadAlloc;
  
  FlushGC(dpy, gc);

  XvGetReq(PutImage, req);

  req->port = port;
  req->drawable = d;
  req->gc = gc->gid;
  req->id = image->id;
  req->src_x = src_x;
  req->src_y = src_y;
  req->src_w = src_w;
  req->src_h = src_h;
  req->drw_x = dest_x;
  req->drw_y = dest_y;
  req->drw_w = dest_w;
  req->drw_h = dest_h;
  req->width = image->width;
  req->height = image->height;
  req->length += (image->data_size + 3) >> 2;

  /* Yes it's kindof lame that we are sending the whole thing,
     but for video all of it may be needed even if displaying
     only a subsection, and I don't want to go through the 
     trouble of creating subregions to send */
  _XSend(dpy, (char *)image->data, image->data_size);

  POSTAMBLE;

  return Success;
}

int XvShmPutImage (
   Display *dpy,
   XvPortID port,
   Drawable d,
   GC gc,
   XvImage *image,
   int src_x,
   int src_y,
   unsigned int src_w,
   unsigned int src_h,
   int dest_x, 
   int dest_y,
   unsigned int dest_w,
   unsigned int dest_h,
   Bool send_event
){
  XShmSegmentInfo *shminfo = (XShmSegmentInfo *)image->obdata;
  xvShmPutImageReq *req;

  PREAMBLE(XvBadExtension);
  
  FlushGC(dpy, gc);

  XvGetReq(ShmPutImage, req);

  req->port = port;
  req->drawable = d;
  req->gc = gc->gid;
  req->shmseg = shminfo->shmseg;
  req->id = image->id;
  req->src_x = src_x;
  req->src_y = src_y;
  req->src_w = src_w;
  req->src_h = src_h;
  req->drw_x = dest_x;
  req->drw_y = dest_y;
  req->drw_w = dest_w;
  req->drw_h = dest_h;
  req->offset = image->data - shminfo->shmaddr;
  req->width = image->width;
  req->height = image->height;
  req->send_event = send_event;

  POSTAMBLE;

  return Success;
}





static Bool
_XvWireToEvent(Display *dpy, XEvent *host, xEvent *wire)
{
  XvEvent *re    = (XvEvent *)host;
  xvEvent *event = (xvEvent *)wire;

  if (event->u.u.type == _XvCodes->first_event+XvVideoNotify)
    {
      re->xvvideo.type = event->u.u.type & 0x7f;
      re->xvvideo.serial = 
	_XSetLastRequestRead(dpy, (xGenericReply *)event);
      re->xvvideo.send_event = ((event->u.u.type & 0x80) != 0);
      re->xvvideo.display = dpy;
      re->xvvideo.time = event->u.videoNotify.time;
      re->xvvideo.reason = event->u.videoNotify.reason;
      re->xvvideo.drawable = event->u.videoNotify.drawable;
      re->xvvideo.port_id = event->u.videoNotify.port;
      if (_XvOldWireToEventVideo) 
	{
	  (void)(* _XvOldWireToEventVideo)(dpy, re, event);
	}
    }
  else if (event->u.u.type == _XvCodes->first_event+XvPortNotify)
    {
      re->xvport.type = event->u.u.type & 0x7f;
      re->xvport.serial = 
	_XSetLastRequestRead(dpy, (xGenericReply *)event);
      re->xvport.send_event = ((event->u.u.type & 0x80) != 0);
      re->xvport.display = dpy;
      re->xvport.time = event->u.portNotify.time;
      re->xvport.port_id = event->u.portNotify.port;
      re->xvport.attribute = event->u.portNotify.attribute;
      re->xvport.value = event->u.portNotify.value;
      if (_XvOldWireToEventPort) 
	{
	  (void)(* _XvOldWireToEventPort)(dpy, re, event);
	}
    }
  else
    {
      return (False);
    }

  return (True);
}

Bool
_XvInitExtension(Display *dpy)
{
  if (!(_XvCodes = XInitExtension (dpy, XvName)))
    {
      return False;
    }

  _XvOldWireToEventVideo = 
    XESetWireToEvent(dpy,
		     _XvCodes->first_event + XvVideoNotify, 
		     _XvWireToEvent);

  _XvOldWireToEventPort = 
    XESetWireToEvent(dpy, 
		     _XvCodes->first_event + XvPortNotify, 
		     _XvWireToEvent);

  _XvOldErrorString = XESetErrorString(dpy, _XvCodes->extension, 
				       _XvErrorString);

  return True;
}

static char *
_XvErrorString(
     Display *dpy,
     int code,
     XExtCodes *ecodes,
     char *buffer,
     int size
){

  if (size < strlen("Encoding") + 1)
    {
      if (_XvOldErrorString) 
	return (* _XvOldErrorString)(dpy, code, ecodes, buffer, size);
    }

  if (code - ecodes->first_error == XvBadPort)
    {
      strcpy(buffer, "Port");
    }
  else if (code - ecodes->first_error == XvBadEncoding)
    {
      strcpy(buffer, "Encoding");
    }
  else
    {
      if (_XvOldErrorString) 
	return (* _XvOldErrorString)(dpy, code, ecodes, buffer, size);
    }

  return (char *)NULL;
}
